builtins: add `Core.throw_methoderror`

topolarity opened the pull request:
This allows us to simulate/mark calls that are known-to-fail.

Required for https://github.com/JuliaLang/julia/pull/54972/

topolarity committed:
builtins: add `Core.throw_methoderror`

This allows us to simulate/mark calls that are known-to-fail.

gbaraldi commented on base/compiler/abstractinterpretation.jl:
```diff
@@ -2226,6 +2226,18 @@ function abstract_throw(interp::AbstractInterpreter, argtypes::Vector{Any}, ::Ab
+    else
+        tmerge(𝕃ᵢ, MethodError, ArgumentError)
+    end
+    return CallMeta(Union{}, exct, EFFECTS_THROWS, NoCallInfo())
```
I guess its not consistent because it captures the world?

topolarity replied:
Great point - I got that right in the tfuncs but not here.

topolarity replied:
Actually, I think this is consistent after all.

Consistency doesn't place requirements on exceptions:
```
   If :consistent functions terminate by throwing an exception, that exception
    itself is not required to meet the egality requirement specified above.
```

And even if it did, it is per-world-age:
```
 The :consistent-cy assertion is made world-age wise. More formally, write
    fᵢ for the evaluation of f in world-age i, then this setting requires:
    ∀ i, x, y: x ≡ y → fᵢ(x) ≡ fᵢ(y)
    However, for two world ages i, j s.t. i ≠ j, we may have fᵢ(x) ≢ fⱼ(y).
```

aviatesk replied:
Yeah, `:consistent` doesn't require the consistency on thrown object. If a method throws consistently, then it is `:consistent` no matter what it throws.

oscardssmith replied:
wait that is a kind of odd spec for it to have. Does that mean we can't inline the throw of an exception if we think a function throws?

aviatesk replied:
Yeah, even if concrete evaluation reveals that a method call will definitely throw, we can't propagate that exception as a `Const`.

aviatesk commented on src/builtins.c:
```diff
@@ -580,6 +580,14 @@ JL_CALLABLE(jl_f_throw)
+{
+    JL_NARGSV(throw_methoderror, 1);
+    size_t world = jl_get_tls_world_age();
+    jl_method_error((jl_function_t *)args[0], &args[1], nargs, world);
```
```suggestion
    jl_method_error((jl_value_t *)args[0], &args[1], nargs, world);
```
since arbitrary object can be a functor?

topolarity committed:
Use `jl_value_t` in place of `jl_function_t`

These are not any different from C's perspective, but one might expect
that a `jl_function_t` is actually `<:Function`
